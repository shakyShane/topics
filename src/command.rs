#[derive(Debug, serde::Deserialize)]
pub struct Command {
    pub cwd: String,
    pub command: String,
    pub title: String,
}
