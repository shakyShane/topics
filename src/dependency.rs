#[derive(Debug, serde::Deserialize)]
pub struct DependencyCheck {
    pub verify: String,
    pub autofix: Option<String>,
    pub url: String,
}
