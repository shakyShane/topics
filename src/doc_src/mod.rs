pub mod doc_src;
pub use doc_src::*;
