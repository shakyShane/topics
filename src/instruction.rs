#[derive(Debug, Clone, serde::Deserialize)]
pub struct Instruction {
    pub instruction: String,
    pub name: String,
}
