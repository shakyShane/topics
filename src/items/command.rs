#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
pub struct Command {
    pub cwd: String,
    pub command: String,
    pub name: String,
}
