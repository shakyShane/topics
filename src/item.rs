use crate::command::Command;
use crate::dependency::DependencyCheck;
use crate::file_exists::FileExistsCheck;
use crate::host::HostEntriesCheck;
use crate::instruction::Instruction;
use crate::topic::Topic;

#[derive(Debug, Clone, serde::Deserialize)]
#[serde(tag = "kind")]
pub enum Item {
    Command(Command),
    FileExistsCheck(FileExistsCheck),
    DependencyCheck(DependencyCheck),
    Instruction(Instruction),
    HostEntriesCheck(HostEntriesCheck),
    Topic(Topic),
}

#[derive(Debug, Clone, serde::Deserialize)]
#[serde(untagged)]
pub enum ItemWrap {
    Named(String),
    Item(Item),
}

impl Item {
    pub fn name(&self) -> String {
        match self {
            Item::Command(cmd) => cmd.name.clone(),
            Item::FileExistsCheck(fec) => fec.name.clone(),
            Item::DependencyCheck(dc) => dc.name.clone(),
            Item::Instruction(inst) => inst.name.clone(),
            Item::HostEntriesCheck(hec) => hec.name.clone(),
            Item::Topic(top) => top.name.clone(),
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_deserialize() -> anyhow::Result<()> {
        let input = r#"
        kind: Topic
        name: run tests
        steps:
            - kind: Instruction
              name: call
              instruction: Call your manager
        "#;
        let t: Topic = serde_yaml::from_str(input)?;
        dbg!(t);
        Ok(())
    }
}
