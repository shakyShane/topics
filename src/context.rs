use crate::doc::{Doc, DocResult};
use crate::opt::Opt;
use crate::print::Print;
use std::path::PathBuf;

#[derive(Debug, Default)]
pub struct Context {
    pub opts: Opt,
}

impl Context {
    pub fn from_opts(opt: &Opt) -> Self {
        Self {
            opts: (*opt).clone(),
        }
    }
    pub fn from_vec(input: &[&str]) -> Self {
        Self::from_opts(&Opt::from_vec(input))
    }
    pub fn join_path(&self, pb: impl Into<PathBuf>) -> PathBuf {
        self.opts.cwd.join_path(pb)
    }
    pub fn read_docs_split(&self) -> (Vec<DocResult<Doc>>, Vec<DocResult<Doc>>) {
        self.read_docs().into_iter().partition(|a| match a {
            Ok(doc) => doc.errors.is_empty(),
            Err(_) => false,
        })
    }
    pub fn read_docs(&self) -> Vec<DocResult<Doc>> {
        self.opts
            .files
            .iter()
            .map(|pb| Doc::from_path_buf(pb, &self))
            .collect()
    }
    pub fn cwd(&self) -> PathBuf {
        self.opts.cwd.0.clone()
    }
}

impl Print for Context {
    fn print(&self, doc: &Doc, ctx: &Context) -> anyhow::Result<()> {
        self.opts.print_kind.print(&doc, &ctx)
    }

    fn print_welcome(&self, docs: &Vec<DocResult<Doc>>, ctx: &Context) -> anyhow::Result<()> {
        self.opts.print_kind.print_welcome(&docs, &ctx)
    }

    fn print_all(&self, docs: &Vec<DocResult<Doc>>, ctx: &Context) -> anyhow::Result<()> {
        self.opts.print_kind.print_all(&docs, &ctx)
    }
    fn print_errors(&self, docs: &Vec<DocResult<Doc>>, ctx: &Context) -> anyhow::Result<()> {
        self.opts.print_kind.print_errors(&docs, &ctx)
    }
}
