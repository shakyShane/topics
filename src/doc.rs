use crate::command::Command;
use crate::context::Context;
use crate::dependency::DependencyCheck;
use crate::doc_src::DocSource;
use crate::instruction::Instruction;
use crate::item::Item;
use crate::topic::Topic;

use std::collections::HashMap;
use std::error::Error;
use std::fmt::{Display, Formatter};
use std::path::PathBuf;
use std::str::FromStr;

#[derive(Debug, Default)]
pub struct Doc {
    pub input_file: PathBuf,
    pub sources: Vec<DocSource>,
    pub topics: HashMap<String, Topic>,
    pub instructions: HashMap<String, Instruction>,
    pub dep_checks: HashMap<String, DependencyCheck>,
    pub commands: HashMap<String, Command>,
}

impl Doc {
    pub fn from_path_buf(pb: &PathBuf, ctx: &Context) -> anyhow::Result<Self> {
        let doc_src = DocSource::from_path_buf(&pb, ctx)?;
        Self::from_doc_src(&pb, doc_src, &ctx)
    }
    pub fn from_doc_src(
        pb: &PathBuf,
        doc_srcs: Vec<DocSource>,
        _ctx: &Context,
    ) -> anyhow::Result<Self> {
        let mut doc = Doc::default();
        doc.input_file = pb.clone();
        doc.sources = doc_srcs;
        for src in &doc.sources {
            let item: Item = serde_yaml::from_str(&src.content).map_err(|e| {
                if let Some(location) = e.location() {
                    DocError::SerdeYamlErr(LocationError {
                        location: Some(Location {
                            line: location.line() + src.line_start,
                            column: location.column(),
                        }),
                        input_file: doc.input_file.clone(),
                        description: e.to_string(),
                    })
                } else {
                    DocError::SerdeYamlErr(LocationError {
                        location: None,
                        input_file: doc.input_file.clone(),
                        description: e.to_string(),
                    })
                }
            })?;
            match item {
                Item::Command(cmd) => {
                    doc.commands.insert(cmd.name.clone(), cmd.clone());
                }
                Item::FileExistsCheck(_) => {}
                Item::DependencyCheck(dc) => {
                    doc.dep_checks.insert(dc.name.clone(), dc.clone());
                }
                Item::Instruction(inst) => {
                    doc.instructions.insert(inst.name.clone(), inst.clone());
                }
                Item::HostEntriesCheck(_) => {}
                Item::Topic(t) => {
                    doc.topics.insert(t.name.clone(), t.clone());
                }
            };
        }
        Ok(doc)
    }
}

#[derive(Debug, thiserror::Error)]
enum DocError {
    #[error(
        "An error occurred when trying to parse a YAML file\n{}",
        .0
    )]
    SerdeYamlErr(LocationError),
}

#[derive(Debug)]
struct LocationError {
    location: Option<Location>,
    input_file: PathBuf,
    description: String,
}

#[derive(Debug)]
struct Location {
    line: usize,
    column: usize,
}

impl Display for LocationError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "");
        writeln!(f, "    msg: `{}`", self.description);
        writeln!(f, "   file: `{}`", self.input_file.display());
        if let Some(location) = &self.location {
            writeln!(f, "   line: `{}`", location.line);
            writeln!(f, " column: `{}`", location.column);
        }
        Ok(())
    }
}

impl FromStr for Doc {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let doc_srcs = DocSource::parse(s)?;
        Doc::from_doc_src(&PathBuf::new(), doc_srcs, &Default::default())
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_single_doc() -> anyhow::Result<()> {
        let input = r#"
kind: Topic
name: Run screen shot tests
deps:
  - global-node
  - global-yarn
steps:
  - github-checkin
"#;
        let doc = Doc::from_str(input)?;
        assert_eq!(doc.sources.len(), 1);
        Ok(())
    }

    #[test]
    fn test_multi_doc() -> anyhow::Result<()> {
        let input = r#"
---
kind: Topic
name: Run screen shot tests
deps:
  - global-node
  - global-yarn
steps:
  - github-checkin
---
---
---
kind: Instruction
instruction: help me!
name: help-me-instruction
---
---
"#;
        let doc = Doc::from_str(input)?;
        assert_eq!(doc.sources.len(), 2);
        Ok(())
    }

    #[test]
    fn test_errors_single() -> anyhow::Result<()> {
        let pb = PathBuf::from("/input-yaml.yml");
        let input = r#"
kind: Topic
name: Run screen shot tests
deps
"#;
        let srcs = DocSource::parse(input)?;
        let doc = Doc::from_doc_src(&pb, srcs, &Default::default());
        insta::assert_debug_snapshot!(doc);
        Ok(())
    }

    #[test]
    fn test_errors_multi() -> anyhow::Result<()> {
        let pb = PathBuf::from("/input-yaml.yml");
        let input = r#"---

kind: DependencyCheck
name: global-node
verify: node -v
url: https://www.nodejs.org

---

kind: DependencyCheck
name: global-yarn
verify: yarn -v
url: https://yarn.sh/legacy

---

kind: Topic
name: Run screen shot tests
deps:
  - global-node
  - global-yarn
steps
"#;
        let srcs = DocSource::parse(input)?;
        let doc = Doc::from_doc_src(&pb, srcs, &Default::default());
        insta::assert_debug_snapshot!(doc);
        Ok(())
    }
}
