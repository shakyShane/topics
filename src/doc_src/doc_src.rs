use crate::context::Context;

use crate::doc::{DocError, DocResult};
use crate::doc_src::DocSourceItems;
use std::path::PathBuf;
use std::str::FromStr;

#[derive(Debug, Default)]
pub struct DocSource {
    pub file_content: String,
    pub doc_src_items: DocSourceItems,
}

impl DocSource {
    pub fn from_path_buf(pb: &PathBuf, ctx: &Context) -> DocResult<Self> {
        let abs = ctx.join_path(pb);
        let file_str = std::fs::read_to_string(&abs).map_err(|e| DocError::PathRead {
            pb: pb.clone(),
            abs: abs.clone(),
            original: e,
        })?;
        let items = DocSourceItems::from_str(&file_str)?;
        let new_self = Self {
            file_content: file_str,
            doc_src_items: items,
        };
        Ok(new_self)
    }
}

impl FromStr for DocSource {
    type Err = DocError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let items = DocSourceItems::from_str(&s)?;
        Ok(Self {
            file_content: s.to_string(),
            doc_src_items: items,
        })
    }
}

#[cfg(test)]
mod test {

    use crate::context::Context;

    use crate::doc_src::DocSource;
    use std::env::current_dir;

    #[test]
    fn test_fixture_file() -> anyhow::Result<()> {
        let ctx = Context::from_vec(&[]);
        let pb = current_dir()?.join("fixtures2/topics.yaml");
        let d = DocSource::from_path_buf(&pb, &ctx)?;
        insta::assert_debug_snapshot!(d);
        Ok(())
    }
}
