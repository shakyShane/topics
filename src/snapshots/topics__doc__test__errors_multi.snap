---
source: src/doc.rs
expression: doc

---
Err(
    SerdeYamlErr(
        LocationError {
            location: Some(
                Location {
                    line: 22,
                    column: 1,
                },
            ),
            input_file: "/input-yaml.yml",
            description: "while parsing a block mapping, did not find expected key at line 7 column 1",
        },
    ),
)
